lib/propolis/src/cpuid.rs

```diff
@@ -0,0 +1,413 @@
+            set.vendor = vendor;
+        }
+        match set.vendor {
+            VendorKind::Amd => {
+                if self.do_cache_topo && self.num_vcpu.is_some() {
+                    self.fix_amd_cache_topo(&mut set);
+                }
+            }
+            _ => {}
+        }
+
+        // apply any requested topo info fixups
+        self.fix_cpu_topo(&mut set);
+
+        // APIC ID based on vcpuid
+        if let Some(vcpuid) = self.vcpuid.as_ref() {
+            if let Some(ent) = set.get_mut(Ident(0x1, None)) {
+                // bits 31:24 contain initial APIC ID
+                ent.ebx &= !0xff000000;
+                ent.ebx |= ((*vcpuid as u32) & 0xff) << 24;
+            }
+        }
+
+        // logical CPU count (if SMT is enabled)
+        if let Some(num_vcpu) = self.num_vcpu.as_ref() {
+            if self.has_smt {
+                if let Some(ent) = set.get_mut(Ident(0x1, None)) {
+                    ent.edx |= (0x1 << 28);
+                    // bits 23:16 contain max IDs for logical CPUs in package
+                    ent.ebx &= !0xff0000;
+                    ent.ebx |= (num_vcpu.get() as u32) << 16;
+                }
+            }
+        }
+
+        set
+    }
+
+    fn fix_amd_cache_topo(&self, set: &mut Set) {
+        assert!(self.do_cache_topo);
+        let num = self.num_vcpu.unwrap().get();
+        for ecx in 0..u32::MAX {
+            match set.get_mut(Ident(0x8000001d, Some(ecx))) {
+                None => break,
+                Some(vals) => {
+                    // bits 7:5 hold the cache level
+                    let visible_count = match (vals.eax & 0b11100000 >> 5) {
+                        0b001 | 0b010 => {
+                            // L1/L2 shared by SMT siblings
+                            if self.has_smt {
+                                2
+                            } else {
+                                1
+                            }
+                        }
+                        0b011 => {
+                            // L3 shared by all vCPUs
+                            // TODO: segregate by sockets, if configured
+                            num as u32
+                        }
+                        _ => {
+                            // unceremonious handling of unexpected cache levels
+                            panic!("unsupported cache level");
+                        }
+                    };
+                    // the number of logical CPUs (minus 1) sharing this cache
+                    // is stored in bits 25:14
+                    vals.eax &= !(0xfff << 14);
+                    vals.eax |= (visible_count - 1) << 14;
+                }
+            }
+        }
+    }
+    fn fix_cpu_topo(&self, set: &mut Set) {
+        for topo in enum_iterator::all::<TopoKind>() {
+            // Nuke any existing info in order to potentially override it
+            let leaf = topo as u32;
+            set.remove_all(leaf);
```

Comment:
This strikes me as a bit odd. If I write a config file that has explicit settings for leaf B/1F/8..1E, but then the specializer declines to pass a value for `num_vcpu`, then the entries I put in the config file will be silently dropped.

Should it be an error to `fix_cpu_topo` for a particular leaf if the original `set` contains that leaf and this function doesn't have enough information to specialize it?

Reply:
What's a bit unfortunate here is that with the builder pattern, we could get the `num_vcpu` after the topology fix-up as been requested.  I'll spell out that the user shouldn't omit num_vcpus in the comment at least.

Reply:
I'm probably missing something, but I'm not sure I follow. IIUC the builder only applies these options when someone calls `execute`, which consumes the whole builder. Before that point I don't think it matters whether you call `with_cpu_topo` or `with_vcpu_count` first, since those are just setting options in the builder, and after that point there's no way to call `with_vcpu_count` anymore because the builder was moved. If all that's right, when would the builder end up getting a CPU count after applying the topo fixups?

Reply:
I was thinking about tossing the error in the `with_cpu_topo` call, but I guess the `execute` call could have a `thiserror` enum covering that case.  I'll take a look at what we might emit there.